# ...................Newsgroups Ciphertext Challenge.........................

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics import classification_report
from sklearn.model_selection import RandomizedSearchCV,GridSearchCV
import warnings
warnings.filterwarnings('ignore')

In [2]:
train= pd.read_csv("C:/Users/Sayan Mondal/Desktop/bmed/train.csv")
test=  pd.read_csv("C:/Users/Sayan Mondal/Desktop/bmed/test.csv")
train.head()

ciphertext            Id  target
0  W)(7yT7h8;Hv\nlTcH;hHhNW)%HhMNPcHNG_Cy^PvMBMy...  ID_48096b7a9      12
1  W)(7yqv(H3cl-iNHv
cNMrcNv3y^](H\ny\tNyc/syPc...  ID_a2d6e55f4       2
2  W)(7yqM()qMl7;HTvM)N\n()8N%hNc8syPcu4MhvyO(H...  ID_9dde342f6      11
3  W)(7yH=j0j?&Il
MHT(HNhhN33cNMrcy^\%7MTyX(cq1%...  ID_b637c2642      16
4  W)(7y8()MHM8lH7v;Nh(7y^q%)\ny8()MHM8/syPcu4Mh...  ID_51b0f15e7       3

In [3]:
train.shape, test.shape

((5631, 3), (13142, 2))

In [4]:
train['target'].value_counts().to_frame().T


15   9    11   8    7    12   2    1    10   13   16   3    4    6   \
target  328  326  312  306  305  300  291  288  288  288  286  283  282  279   

         17   14   5    0    18   19  
target  276  275  274  244  219  181

In [6]:
train['ciphertext'] = train['ciphertext'].str.lower()
train['ciphertext'] = train['ciphertext'].astype(str)

In [7]:
test['ciphertext'] = test['ciphertext'].str.lower()
test['ciphertext'] = test['ciphertext'].astype(str)


In [8]:
train.head(3)


ciphertext            Id  target
0  w)(7yt7h8;hv\nltch;hhhnw)%hhmnpchng_cy^pvmbmy...  ID_48096b7a9      12
1  w)(7yqv(h3cl-inhv
cnmrcnv3y^](h\ny\tnyc/sypc...  ID_a2d6e55f4       2
2  w)(7yqm()qml7;htvm)n\n()8n%hnc8sypcu4mhvyo(h...  ID_9dde342f6      11

In [57]:
Ciph = train.ciphertext


In [58]:
import re
from nltk.corpus import stopwords
stop1 = stopwords.words("english")


### TF_IDF Word and Char level features

In [60]:
token= ((u'(?ui)\\b\\w*[a-z]+\\w*\\b'))

## word....
word_vec = TfidfVectorizer(
     sublinear_tf=True,
     strip_accents='ascii',
     analyzer='word',
     token_pattern=token,
     dtype=np.float32,
     max_features=9100)
     

### Character .....
char_vec = TfidfVectorizer(
     sublinear_tf=True,
     strip_accents='ascii',
     analyzer='char',
     token_pattern=token,
     dtype=np.float32,
     max_features=10000)
     
    



In [14]:
X1=word_vec.fit_transform(Ciph)


In [15]:
print(word_vec.get_feature_names())

['00n', '00ni', '00y', '06i', '06j', '06u', '0a', '0c', '0e', '0f', '0g', '0i', '0i0', '0i6', '0ij', '0in', '0ini', '0inj', '0j', '0j0', '0j6i', '0ji', '0jj', '0jn', '0jni', '0jnj', '0jy', '0l', '0l1', '0m', '0n', '0n0', '0n6', '0n_', '0ni', '0nj', '0nlb7h7tnhtc', '0nn', '0nni', '0nsy', '0ny', '0nyy', '0o', '0p', '0q', '0r', '0s', '0sy', '0sy_', '0syc', '0syoo', '0sysy', '0sysysy', '0syx', '0u', '0w', '0x', '0y', '0y3', '0yb', '0yh', '0yq', '0yr', '0yu', '0yy', '0z', '116pv', '11c7', '11ct', '11ctv', '11cv', '11l', '11l78t', '11m', '11m1y', '11mh', '11mhqmy', '11mhqmyv', '11mhv', '11mhvl', '11mhvy', '11mhy', '11mq', '11mqm', '11mqmsyp', '11mqmy', '11mqmym', '11mqnmrc', '11mr', '11mrsy', '11mry', '11mry3', '11mry7', '11mryc', '11mrye', '11mryl', '11mryu', '11mryv', '11my', '11nh', '11nmrc', '11ns', '11nsy', '11nsysy', '11ny', '11nyy', '11s', '11sy', '11syh', '11syl', '11syp', '11syumy', '11syv', '11t', '11tsy', '11ty', '11tyl', '11tyv', '11u', '11v', '11y', '11y1', '11y1m', '11y3', '11y

In [17]:
print(word_vec.vocabulary_)


{'w': 7862, 'hv': 3337, 'hhmhvy': 2431, 'sypcu4mhv': 6340, 'y': 7976, 'm': 4189, 'vyr': 7802, 'yoch1m': 8520, 'yp': 8540, 'hqy': 2884, '3m': 633, 'tyr': 7030, 'sy_': 6272, 'q': 5694, 'h': 2191, 'i': 3656, 'v': 7346, 'hmt': 2603, 'y0': 7977, 'syx': 6413, 'tv': 6803, 'ucv': 7216, 'y3': 8017, '1rsyoo': 393, 'm6m': 4293, 'hq6': 2791, 'r': 5910, 'nl': 5349, 'rt': 6117, 'sysy': 6351, 'hy': 3430, 'er': 2058, 'n': 5249, 'yt3': 8712, '18m': 187, 'ij': 3690, 'h6': 2237, 'j6': 3783, 'vmt': 7485, 'sy': 6249, 'ysy': 8674, 'yr': 8637, 'huvy8h': 3311, '3y3': 684, 'myv': 5222, 'tvyv': 6919, 'tykcmtv': 6997, 'hyt': 3589, 'yl': 8402, 'qc': 5712, 'mryv': 4710, 'vsy': 7574, 'yv': 8861, 'tyu': 7070, 'ry3': 6149, 'c1ryumy7': 1583, 'tvy': 6888, 'vmnsy': 7445, 'ty3': 6932, 'hrm': 3000, 'u': 7103, 'cvyv': 1949, 'tmy7': 6655, 'tt': 6747, 'bmyh': 1455, 'hh': 2404, 'tyv': 7089, 'mymbm': 5085, 'mtmhvy': 4795, 'yt': 8708, 'vmtny': 7487, 'y1': 7996, '8y1': 1219, '8myh': 1171, 'tsy': 6725, 'vy': 7705, 'bmyummhy': 148

In [18]:
X2=char_vec.fit_transform(Ciph)

In [19]:
print(char_vec.get_feature_names())

['\x02', '\x03', '\x06', '\x08', '\t', '\n', '\x10', '\x18', '\x1a', '\x1b', '\x1c', '\x1e', ' ', '!', '"', '#', '$', '%', '&', '(', ')', '*', '+', ',', '-', '.', '/', '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ':', ';', '<', '=', '>', '?', '@', '[', '\\', ']', '^', '_', '`', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '{', '|', '}', '~', '\x7f']


In [20]:
print(char_vec.vocabulary_)

{'w': 72, ')': 20, '(': 19, '7': 34, '\x1a': 8, 'y': 74, 't': 69, 'h': 57, '8': 35, ';': 38, 'v': 71, '\n': 5, 'l': 61, 'c': 52, 'n': 63, '%': 17, 'm': 62, 'p': 65, 'g': 56, '_': 48, '^': 47, 'b': 51, '9': 36, '6': 33, ':': 37, 'o': 64, '/': 26, 's': 68, 'u': 70, '4': 31, '}': 78, '\x03': 1, '\x1e': 11, 'r': 67, '1': 28, 'q': 66, ']': 46, '3': 30, '|': 77, 'i': 58, '0': 27, '+': 22, 'x': 73, '\x1b': 9, 'f': 55, 'k': 60, '=': 40, 'e': 54, '"': 14, '-': 24, '@': 43, '\x1c': 10, '`': 49, 'j': 59, '&': 18, ' ': 12, '?': 42, '\x02': 0, ',': 23, 'a': 50, '\x10': 6, '\t': 4, '2': 29, 'z': 75, '\\': 45, '5': 32, '<': 39, '[': 44, '*': 21, '>': 41, '#': 15, '~': 79, '.': 25, '{': 76, '\x7f': 80, 'd': 53, '!': 13, '\x18': 7, '\x06': 2, '\x08': 3, '$': 16}


### Stacking the TF-IDF word and char level features.

In [21]:
from scipy.sparse import hstack, csr_matrix #need this if you want to save tfidf_matrix

In [22]:
X = hstack([X1,X2])
y=train.target

In [23]:
X.shape

(5631, 9181)

In [24]:
X.shape,y.shape

((5631, 9181), (5631,))

### Train-Test Split

In [25]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)

### Logistic Regression..



In [26]:
from sklearn.linear_model import LogisticRegression

In [27]:
lr=LogisticRegression()

In [28]:
lr= LogisticRegression()
                   
lr.fit(X,y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='lbfgs', tol=0.0001, verbose=0,
                   warm_start=False)

In [29]:
lr_pred=lr.predict(X_test)

In [30]:
print(classification_report(y_test,lr_pred)) 

              precision    recall  f1-score   support

           0       0.95      0.95      0.95        55
           1       0.82      0.87      0.85        54
           2       0.98      0.91      0.95        68
           3       0.93      0.96      0.95        57
           4       0.97      1.00      0.98        57
           5       0.90      0.98      0.94        57
           6       0.88      0.92      0.90        49
           7       0.94      0.94      0.94        66
           8       0.92      0.95      0.94        62
           9       0.97      0.98      0.97        59
          10       0.97      1.00      0.98        59
          11       0.95      0.92      0.93        62
          12       0.93      0.89      0.91        57
          13       0.97      0.98      0.97        58
          14       0.92      0.94      0.93        47
          15       0.92      1.00      0.96        65
          16       0.89      0.96      0.93        52
          17       1.00    

In [31]:
from sklearn.metrics import cohen_kappa_score
cohen_kappa_score(y_test,lr_pred)

0.933617170253205

### HyperParameter Optinization of Logistic regression using GridsearchCV

In [32]:
%%time
lr = LogisticRegression()
solvers = ['newton-cg', 'lbfgs', 'sag'] 
penalty = ['l2']  #he ‘newton-cg’, ‘sag’ and ‘lbfgs’ solvers support only l2 penalties
c_values = [ 10, 1.0, 0.1, 0.01] 
# define grid search
grid = dict(solver=solvers,penalty=penalty,C=c_values)
grid_search = GridSearchCV(estimator=lr, param_grid=grid, n_jobs=-1, cv=5, scoring='accuracy',error_score=0)
grid_result = grid_search.fit(X, y)

Wall time: 5min 34s


In [33]:
# summarize results
print("Grid_accuracy: %f using %s" % (grid_result.best_score_, grid_result.best_params_))

Grid_accuracy: 0.756704 using {'C': 10, 'penalty': 'l2', 'solver': 'newton-cg'}


In [34]:
# finding best estimator..#
grid_result.best_estimator_

LogisticRegression(C=10, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='newton-cg', tol=0.0001, verbose=0,
                   warm_start=False)

In [35]:
lr= LogisticRegression(C=10, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='newton-cg', tol=0.0001, verbose=0,
                   warm_start=False)
                   
lr.fit(X,y)


LogisticRegression(C=10, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l2',
                   random_state=None, solver='newton-cg', tol=0.0001, verbose=0,
                   warm_start=False)

In [36]:
lr_pred=lr.predict(X_test)

In [37]:
print(classification_report(y_test,lr_pred)) 

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        55
           1       1.00      1.00      1.00        54
           2       1.00      1.00      1.00        68
           3       1.00      1.00      1.00        57
           4       1.00      1.00      1.00        57
           5       1.00      1.00      1.00        57
           6       1.00      0.98      0.99        49
           7       1.00      1.00      1.00        66
           8       1.00      1.00      1.00        62
           9       1.00      1.00      1.00        59
          10       1.00      1.00      1.00        59
          11       0.98      1.00      0.99        62
          12       0.98      1.00      0.99        57
          13       1.00      1.00      1.00        58
          14       1.00      1.00      1.00        47
          15       1.00      1.00      1.00        65
          16       1.00      0.98      0.99        52
          17       1.00    

So we can see, by changing few parameters like 'C', 'solver' model performence improve drastically.

In [35]:
from sklearn.metrics import cohen_kappa_score
cohen_kappa_score(y_test,lr_pred)

0.9990646907047518

In kappa Scale 0.81–1.00 considered as almost perfect agreement,
our model score is 0.99906.

### Applying the test data on Logistic Regression. As it provides the best Classification result

In [38]:
X3= test.ciphertext ## taking ciphertext from test data

In [39]:
X31=word_vec.transform(X3)

In [40]:
X32=char_vec.transform(X3)

In [42]:
X33 = hstack([X31,X32])

In [43]:
X33.shape

(13142, 9181)

In [44]:
lr_pred_1=lr.predict(X33)

In [46]:
df = pd.DataFrame({'Id':test['Id'] ,'Predicted': lr_pred_1})

In [49]:
df.head(10)

Id  Predicted
0  ID_e93d1d4c6          2
1  ID_f5f7560ec          2
2  ID_6ebe8f07f          9
3  ID_26222d9b7         18
4  ID_a0632653b          6
5  ID_641f090da         14
6  ID_888b6c7a5         12
7  ID_7b5b3bb47         17
8  ID_70f08430c         11
9  ID_2562b050a          6

In [48]:
fin_result=df.to_csv("C:/Users/Sayan Mondal/Desktop/bmed/submission.csv", index=False)

### Naive Bayes Classifier

In [50]:
from sklearn.naive_bayes import MultinomialNB
nb=MultinomialNB()

In [51]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

In [52]:
model=nb.fit(X_train, y_train)

In [53]:
y_pred = model.predict(X_test)

In [54]:
print(classification_report(y_test,y_pred)) 

              precision    recall  f1-score   support

           0       1.00      0.18      0.30        45
           1       0.71      0.18      0.29        56
           2       0.67      0.12      0.20        68
           3       0.58      0.12      0.20        58
           4       1.00      0.05      0.10        60
           5       0.83      0.19      0.31        53
           6       0.86      0.35      0.49        52
           7       0.12      0.86      0.20        44
           8       0.44      0.58      0.50        57
           9       0.37      0.90      0.53        63
          10       0.80      0.18      0.30        65
          11       0.94      0.69      0.79        70
          12       0.64      0.43      0.52        58
          13       1.00      0.12      0.21        68
          14       1.00      0.17      0.29        65
          15       0.20      1.00      0.33        60
          16       1.00      0.41      0.58        61
          17       0.97    

In [55]:
from sklearn.metrics import cohen_kappa_score
cohen_kappa_score(y_test,y_pred)

0.3346821000687218

### Random Forest Classifier

In [29]:
from sklearn.ensemble import RandomForestClassifier

rf= RandomForestClassifier(n_estimators=200,n_jobs=-1,)

In [30]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
X_train.shape,y_train.shape


((4504, 9081), (4504,))

In [31]:
model=rf.fit(X_train, y_train)

y_pred = model.predict(X_test)
  

In [32]:
print(classification_report(y_test,y_pred)) 

              precision    recall  f1-score   support

           0       0.74      0.58      0.65        59
           1       0.43      0.28      0.34        58
           2       0.47      0.68      0.56        59
           3       0.59      0.45      0.51        60
           4       0.38      0.35      0.37        57
           5       0.60      0.69      0.64        55
           6       0.34      0.75      0.46        48
           7       0.44      0.31      0.36        68
           8       0.51      0.54      0.53        72
           9       0.56      0.70      0.62        64
          10       0.70      0.75      0.72        52
          11       0.75      0.76      0.76        59
          12       0.40      0.47      0.43        51
          13       0.77      0.53      0.63        64
          14       0.68      0.60      0.64        47
          15       0.50      0.86      0.63        59
          16       0.74      0.55      0.63        67
          17       0.84    

In [33]:
cohen_kappa_score(y_test,y_pred)

0.5380358028932403

## CatBoost Classifier

In [99]:
from catboost import CatBoostClassifier

In [96]:
cb=CatBoostClassifier(iterations=150,learning_rate= 0.20)
cb.fit(X_train,y_train)

0:	learn: 2.9007626	total: 26.9s	remaining: 1h 6m 46s
1:	learn: 2.8386856	total: 54.8s	remaining: 1h 7m 36s
2:	learn: 2.7880172	total: 1m 22s	remaining: 1h 7m 34s
3:	learn: 2.7432547	total: 1m 56s	remaining: 1h 10m 44s
4:	learn: 2.7199767	total: 2m 28s	remaining: 1h 11m 49s
5:	learn: 2.6804945	total: 3m 2s	remaining: 1h 13m 4s
6:	learn: 2.6568111	total: 3m 33s	remaining: 1h 12m 35s
7:	learn: 2.6281853	total: 4m 3s	remaining: 1h 11m 55s
8:	learn: 2.6117318	total: 4m 32s	remaining: 1h 11m 10s
9:	learn: 2.5838894	total: 5m 2s	remaining: 1h 10m 31s
10:	learn: 2.5688622	total: 5m 33s	remaining: 1h 10m 19s
11:	learn: 2.5522563	total: 6m 10s	remaining: 1h 11m 6s
12:	learn: 2.5178514	total: 6m 40s	remaining: 1h 10m 17s
13:	learn: 2.5092027	total: 7m 10s	remaining: 1h 9m 38s
14:	learn: 2.4942745	total: 7m 39s	remaining: 1h 8m 55s
15:	learn: 2.4803577	total: 8m 9s	remaining: 1h 8m 15s
16:	learn: 2.4599820	total: 8m 37s	remaining: 1h 7m 31s
17:	learn: 2.4442231	total: 9m 7s	remaining: 1h 6m 54s
1

In [97]:
cb_pred = cb.predict(X_test)
print(classification_report(y_test,cb_pred )) 

              precision    recall  f1-score   support

           0       0.59      0.77      0.67        47
           1       0.33      0.33      0.33        52
           2       0.67      0.56      0.61        63
           3       0.54      0.35      0.43        54
           4       0.33      0.40      0.36        50
           5       0.42      0.89      0.57        37
           6       0.49      0.72      0.58        53
           7       0.35      0.50      0.41        56
           8       0.49      0.51      0.50        72
           9       0.70      0.60      0.65        72
          10       0.67      0.68      0.67        71
          11       0.88      0.69      0.77        71
          12       0.54      0.41      0.47        66
          13       0.58      0.66      0.62        56
          14       0.80      0.58      0.67        55
          15       0.75      0.83      0.79        66
          16       0.56      0.50      0.53        56
          17       0.86    

In [98]:
cohen_kappa_score(y_test,cb_pred )

0.5511830230100008

So, Considering the all above model performence, Finalized model is Logistic Regression. And i have applied test data on the same.